### PR TITLE
API-2831 fix nested nots

### PR DESCRIPTION
**Pull request by GutsyGusty**
@wlmcewen and @joshcom Mind taking a look at this?

### CODE REVIEW

File: test/unit/parser_test.rb [hunk 1]
```diff
@@ -24,7 +24,8 @@ def test_conjunction
     expression = @parser.parse('Test Eq 10 Not Test Ne 11')
     assert_equal 10.to_s, expression.first[:value]
     assert_equal 11.to_s, expression.last[:value]
-    assert_equal 'Not', expression.last[:conjunction]
+    assert_equal 'And', expression.last[:conjunction]
```
Comment (wlmcewen) [inline test/unit/parser_test.rb]: @joshcom, this is perhaps the case you'll be most interested in based on your recent work.

File: lib/sparkql/parser_tools.rb [hunk 2]
```diff
@@ -42,12 +42,31 @@ def tokenize_conjunction(exp1, conj, exp2)
...
+    exp2.first[:conjunction_level] = @lexer.level
+
+    flip_expression_if_necessary(exp2, conj)
+
+    exp1 + exp2
+  end
+
+  def tokenize_unary(conj, exp)
+    flip_expression_if_necessary(exp, conj)
     exp.first[:unary_level] = @lexer.level
```
Comment (wlmcewen) [inline lib/sparkql/parser_tools.rb]: This sounds right... only use the unary_level that matched the first not. However, this willl stick around even when we flip the not off. I don't think it's used in any case.

**Comment by wlmcewen** [comment]
Looks good to me, we'll need to be careful when it comes to integration, but it's the right track long term.

File: lib/sparkql/parser_tools.rb [hunk 3]
```diff
@@ -42,12 +42,31 @@ def tokenize_conjunction(exp1, conj, exp2)
...
     exp
   end
-    
+
+  def flip_expression_if_necessary(exp, conj)
+    if exp.first[:unary].nil?
+      exp.first[:unary] = conj
+    else
+      exp.first.delete(:unary)
```
Comment (GutsyGusty) [inline lib/sparkql/parser_tools.rb]: Wade mentioned this.. cannot use `first` here. Will modify and write test for that.

File: test/unit/parser_test.rb [hunk 4]
```diff
@@ -408,6 +410,58 @@ def test_expression_conditions_attribute
...
+    assert_equal "1000", exp[:value]
+    assert_nil exp[:unary]
+
+    exp = expressions[1]
+    assert_equal "And", exp[:conjunction]
+    assert_equal "2000", exp[:value]
+    assert_nil exp[:unary]
+
+    @parser = Parser.new
+    expressions = @parser.parse('Not (Not (Not ListPrice Eq 1000 Not ListPrice Eq 2000))')
```
Comment (joshcom) [inline test/unit/parser_test.rb]: I'd like to see a test (I didn't catch one in the existing set) that ensures something like `Not (ListPrice Eq 1000 And City Eq 'Fargo')` still works, now that `Not` is always a conjunction.

Comment (GutsyGusty) [inline test/unit/parser_test.rb]: Added the test, but it would be more accurate to say that `Not` is *never* a conjunction

**Pull request closed without merging**